Merge Android archives

djee-ms opened the pull request:
Merge the two Android archives (`*.aar`) previously needed for Android support in Unity into a single archive `mrwebrtc.aar`. This drastically reduces the storage footprint of a Unity Android project by shipping a single archive with a single version of `libwebrtc`.

Bugs: #247, #248

djee-ms commented:
This is missing a way to import `UnityUtility.java` which is not compiled in `libwebrtc.aar` by default. I had a local change, need to find a way around it by _e.g._ having our own utility class.

fibann commented on libs/unity/library/Runtime/Scripts/Media/MediaSender.cs:
```diff
@@ -126,6 +126,12 @@ public void Unmute()
         /// <inheritdoc/>
         protected async Task OnEnable()
         {
+#if PLATFORM_ANDROID
```
If we #if out the whole `Android` class we then have to #if out all the places of call. We should #if out the method bodies instead, and possibly rename `Android` to `Platform` or something like this.

djee-ms replied:
I appreciate the first part, which is what Unity does I think; I guess you can `using Unity.Android` on Windows and it does nothing. On the second part I disagree because this is really Android-specific.

djee-ms replied:
Fixed.

fibann commented on tools/build/android/webrtc-native/build.gradle:
```diff
@@ -66,18 +66,27 @@ dependencies {
-    if (task.name == 'externalNativeBuildRelease') {
+    // Once build is done, merge the AAR archives
+    if (task.name == 'bundleReleaseAar') {
         task.finalizedBy copyToUnitySample
-    } else if (task.name == 'externalNativeBuildDebug') {
+        copyToUnitySample.dependsOn(mergeLibwebrtcArchive)
+    } else if (task.name == 'bundleDebugAar') {
         task.finalizedBy copyToUnitySample
-    } else if (task.name == 'externalNativeBuildFulldebug') {
+        copyToUnitySample.dependsOn(mergeLibwebrtcArchive)
+    } else if (task.name == 'bundleFulldebugAar') {
         task.finalizedBy copyToUnitySample
+        copyToUnitySample.dependsOn(mergeLibwebrtcArchive)
```
I don't know the gradle syntax, but it seems to me that the body of all 3 branches is the same 😄 Merge?

djee-ms replied:
Yeah I was lazy, didn't want to figure out Gradle. I guess some `task.name.beginsWith("externalNativeBuild")` would be best, if that's possible. I'll have a look at the Gradle docs.

djee-ms replied:
Hum.. actually I was looking at the old code. I will just add a single `if`.

djee-ms replied:
Fixed.

djee-ms commented:
Fixed with rebase to keep separate commits for individual works. Changes since last iteration are:
- Moved `#if UNITY_ANDROID` inside `Android.Initialize()` and removed it from other places.
- Fixed the tasks in Gradle (last 3 blocks in `build.gradle`)

fibann commented on tools/build/android/webrtc-native/src/main/cpp/jni_onload.cpp:
```diff
@@ -2,25 +2,91 @@
+  // webrtc::jni::LoadGlobalClassReferenceHolder(); //deprecated
+  webrtc::InitClassLoader(webrtc::jni::GetEnv());
+
+  // This is called by the official JNI_OnLoad() of
+  // examples/androidnativeapi/jni/onload.cc. This seems to initialize a
+  // completely different JVM/JNI set of C++ references
+  // (modules/utility/source/jvm_android.cc), only used by the audio module.
+  // Apparently it seems this was to become the "new way" but got cancelled and
+  // the change partially reverted.
+  // https://bugs.chromium.org/p/webrtc/issues/detail?id=8067
+  // Now, in addition, when intializing WebRTC from Java/Unity, this is actually
+  // automatically called from
+  // JNI_PeerConnectionFactory_InitializeAndroidGlobals() which is called from
+  // Java via the Unity wrapper code in Android.Initialize(), so cannot be
+  // called manually here otherwise it asserts.
```
Very nice!
